File: Frontend/nimrag-frontend/src/types/commands.ts
import type { InputActionConfig } from './interactions'

export type CommandModalityType = 'gesture' | 'voice' | 'musical_audio' | 'keyboard' | 'dev'

export interface CommandModalitySettings {
  enabled: boolean
  active_training_artifact_id: string | null
  metadata: Record<string, unknown>
}

export interface CommandDevicePreferences {
  gesture_camera_index: number | null
  voice_device_index: number | null
  musical_audio_device_index: number | null
}

export interface CommandProfile {
  profile_id: string
  display_name: string
  description: string | null
  input_action_config: InputActionConfig
  modality_settings: Record<CommandModalityType, CommandModalitySettings>
  device_preferences: CommandDevicePreferences
  updated_at: string | null
}

export interface CommandProfilesConfig {
  active_profile_id: string
  profiles: CommandProfile[]
  updated_at: string | null
}

export interface CommandProfilesConfigEnvelope {
  config: CommandProfilesConfig
}

export interface MusicalAudioConfig {
  enabled: boolean
  device_index: number
  sample_rate: number
  block_size: number
  queue_max_chunks: number
  silence_threshold: number
  pitch_confidence_threshold: number
  command_cooldown_seconds: number
  min_pattern_notes: number
  max_pattern_window_seconds: number
  active_artifact_id: string | null
  updated_at: string | null
}

export interface MusicalAudioConfigEnvelope {
  config: MusicalAudioConfig
}

export interface MusicalAudioNoteEvent {
  relative_pitch_semitones: number
  relative_time_seconds: number
  duration_seconds: number | null
  confidence: number | null
}

export interface MusicalAudioTrainingArtifact {
  artifact_id: string
  profile_id: string
  raw_input: string
  display_name: string
  source_hint: 'whistle' | 'voice' | 'instrument' | 'unknown'
  notes: MusicalAudioNoteEvent[]
  match_threshold: number
  minimum_confidence: number
  sample_count: number
  enabled: boolean
  created_at: string | null
  updated_at: string | null
  metadata: Record<string, unknown>
}

export interface MusicalAudioTrainingArtifactEnvelope {
  artifact: MusicalAudioTrainingArtifact
}

export interface MusicalAudioTrainingArtifactListEnvelope {
  artifacts: MusicalAudioTrainingArtifact[]
}

File: Backend/src/schemas/commands.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, model_validator

from schemas.interactions import InputActionConfig

CommandModalityType = Literal["gesture", "voice", "musical_audio", "keyboard", "dev"]


class CommandModalitySettings(BaseModel):
    enabled: bool = True
    active_training_artifact_id: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


def build_default_modality_settings() -> (
    dict[CommandModalityType, CommandModalitySettings]
):
    return {
        "gesture": CommandModalitySettings(enabled=True),
        "voice": CommandModalitySettings(enabled=True),
        "musical_audio": CommandModalitySettings(enabled=False),
        "keyboard": CommandModalitySettings(enabled=True),
        "dev": CommandModalitySettings(enabled=True),
    }


class CommandDevicePreferences(BaseModel):
    gesture_camera_index: int | None = Field(default=None, ge=0)
    voice_device_index: int | None = Field(default=None, ge=-1)
    musical_audio_device_index: int | None = Field(default=None, ge=-1)


class CommandProfile(BaseModel):
    profile_id: str = Field(default="default", min_length=1)
    display_name: str = Field(default="Standard", min_length=1)
    description: str | None = None
    input_action_config: InputActionConfig = Field(default_factory=InputActionConfig)
    modality_settings: dict[CommandModalityType, CommandModalitySettings] = Field(
        default_factory=build_default_modality_settings
    )
    device_preferences: CommandDevicePreferences = Field(
        default_factory=CommandDevicePreferences
    )
    updated_at: datetime | None = None


class CommandProfilesConfig(BaseModel):
    active_profile_id: str = Field(default="default", min_length=1)
    profiles: list[CommandProfile] = Field(default_factory=lambda: [CommandProfile()])
    updated_at: datetime | None = None

    @model_validator(mode="after")
    def ensure_active_profile_exists(self) -> "CommandProfilesConfig":
        if not self.profiles:
            self.profiles = [CommandProfile()]

        if not any(
            profile.profile_id == self.active_profile_id for profile in self.profiles
        ):
            self.active_profile_id = self.profiles[0].profile_id

        return self


class CommandProfilesConfigEnvelope(BaseModel):
    config: CommandProfilesConfig
